### ISSUE TITLE
allow excluding certain files being copied to the build folder

### ISSUE THREAD

**Issue by haochenx** [ocaml/dune#4865]
## Desired Behavior

Currently there are ways to exclude certain directories from being copied to the build folder, but there is no way (or at least I could not find one reading the Stanza Reference page) to stop copying certain files.

I would like to suggest adding a stanza like `(ignored_files ...)` or a stanza like `(files ...)` to allow explicit exclusion of files to be copied to the build folder.

I believe it is in general nice to have this possible, and it comes especially handy when the build needs to execute external commands whose behavior may depends on specific files in the current directory. For example, in my case I would like to exclude the `.git` file under a git submodule, whose presence in the build folder messes up invocations to `git`, and I am using a workaround to delete the copied `.git` file before invoking any `git` commands.

#### project structure
```
<project_root>/
- .some-file-to-be-excluded
- dune
- dune-project
- module1/
  - ...
```

#### `_build` directory with an empty `dune` file
```
_build/default
- .some-file-to-be-excluded
- dune
- dune-project
- module1/
  - ...
```

#### contents of `dune`, `(ignore_files ...)` variant
```
(ignored_files .some-file-to-be-excluded)
```

#### contents of `dune`, `(files ...)` variant
```
(files :standard \ .some-file-to-be-excluded)
```

#### `_build` directory with a `dune` file of either variant listed above
```
_build/default
- dune
- dune-project
- module1/
  - ...
```

**Comment by nojb** [comment]
By default any directory starting with `.` or `_` will be ignored by `dune` (and in particular, it will not be copied to the `_build` directory). If this is not the case, then it sounds rather like a bug in `dune`.

Can you try to come up with a minimal reproduction of this issue?

More generally you can control which directories are ignored by `dune` by using the `(dirs ...)` stanza.

**Comment by nojb** [comment]
OK, sorry I noticed you asked about *files*, and not *directories*. I got confused because I had never heard about this `.git` *file* associated to git submodules. Do you have a pointer to some documentation about this file?

Also, note that on 3.0 the behaviour of `dune` will be changed so that only files which are dependencies of the targets being built will be copied to `_build` (see #4461).

**Comment by haochenx** [comment]
i could confirm that if `.git` is a directory, it gets ignored by default. but in my case, since the dune project is included in a larger repository as a git submodule, it is a `.git` file with contents like `gitdir: ../../.git/modules/submodule_dir`

**Comment by haochenx** [comment]
>I had never heard about this .git file associated to git submodules. Do you have a pointer to some documentation about this file?

well, I believe the official reference would be [this git manual page]. i will try to prepare a minimal example and upload it here.

[this git manual page]: https://git-scm.com/docs/git-submodule

**Comment by haochenx** [comment]
> Also, note that on 3.0 the behaviour of dune will be changed so that only files which are dependencies of the targets being built will be copied to _build (see #4461).

it's nice to hear about that, which make it sounds to me like that my problem with `.git` file should disappear on 3.0. 

is there an estimated date for the release of 3.0? if it's coming out soon, maybe we could just use workarounds like the one mentioned in the original issue.

**Comment by nojb** [comment]
> well, I believe the official reference would be [this git manual page](https://git-scm.com/docs/git-submodule). i will try to prepare a minimal example and upload it here.

Right, thanks, I don't think the minimal example is worth the trouble, the issue is clear.

> is there an estimated date for the release of 3.0? if it's coming out soon, maybe we could just use workarounds like the one mentioned in the original issue.

As far as I know it should happen in the next couple of months (see https://dune.build/blog/dune-3-plan). I agree with you that using a workaround like the one you already have is probably the best at this point.

**Comment by haochenx** [comment]
i just got a minimal example: [dune_4865.ba33a.zip](https://github.com/ocaml/dune/files/7019620/dune_4865.ba33a.zip)


if you go to `<root>/submod` and run `dune build`, you'd get error message like the following
```
         git version.ml (exit 128)
(cd _build/default && /usr/local/bin/git rev-parse --short HEAD) > /var/folders/hm/wr72kh555mb9r0mpl_fj6pp00000gn/T/build4c7b80.dune/dune-pipe-action-6e8bac.stdout
fatal: not a git repository (or any of the parent directories): .git
```

but if you copy `<root>/submod` some where else and delete the `.git` file there (or uncomment the workaround in `<root>/submod/dune`), you could build the project without problem.


> As far as I know it should happen in the next couple of months (see https://dune.build/blog/dune-3-plan). I agree with you that using a workaround like the one you already have is probably the best at this point.

i see. in that case we will be using this workaround and see whether migrating to 3.0 would fix the problem after it get released.

**Comment by rgrinberg** [comment]
The feature request is still reasonable. I don't see why we shouldn't have `files` that's analogous to `dirs`. This still low priority, so unless you want to tackle this yourself, it's unlikely to get done. If you need some pointers, feel free to ask and we can show you what needs to be modified.
